### PR TITLE
BAH-4832 | Transformation of Observation Form data from/ to FHIR format to be taken care in form2-controls library (refactor)

### PR DESCRIPTION
**Pull request by sahupadma**
JIRA - https://bahmni.atlassian.net/browse/BAH-4382

 Summary                                                                                                                                                                      
                                                                                                                                                                               
  Introduces a FHIR Observation transformer to convert form2 observations into FHIR R4-compliant Observation resources, enabling seamless integration with FHIR-based APIs.

 Key Features                                                                                                                                                                 
                                                                                                                                                                               
  - Value Type Support: Handles numeric, string, boolean, date/datetime, coded values, and complex datatypes                                                                   
  - Group Observations: Recursively processes groupMembers with proper hasMember FHIR references                                                                               
  - Extensions: Preserves form namespace/field path and complex data attachments via FHIR extensions                                                                           
  - Interpretation Mapping: Converts ABNORMAL/NORMAL to standard FHIR interpretation codes                                                                                     
  - Clinical Notes: Maps observation comments to FHIR notes                                                                                                                    
  - Voided Filtering: Automatically excludes voided observations                                                                                                                                               
                                                                                                                                                                               
  Exports:                                                                                                                                                                     
  - FhirObservationTransformer class and all FHIR constants exported from src/index.jsx                                                                                        
  - TypeScript definitions added to index.d.ts for FhirReference, FhirTransformOptions, FhirObservationEntry, and the transformer class                                        
                                                                                                                                                                               
  Tests:                                                                                                                                                                       
  - Comprehensive test suite with 545 lines covering all transformation scenarios, edge cases, and group member handling

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -67,8 +68,7 @@
     "style-loader": "^3.3.3",
     "webpack": "^5.89.0",
     "webpack-cli": "^5.1.4",
-    "webpack-dev-server": "^4.15.1",
-    "@fortawesome/fontawesome-free": "^7.1.0"
+    "webpack-dev-server": "^4.15.1"
```
Comment (sumaztwcode) [inline package.json:71]: need for this? this won't go into prod rt?

File: src/helpers/FhirObservationTransformer.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+  CONCEPT_DATATYPE_NUMERIC,
+  CONCEPT_DATATYPE_COMPLEX,
+  FHIR_OBSERVATION_STATUS_FINAL,
+  FHIR_RESOURCE_TYPE_OBSERVATION,
+  DATE_REGEX_PATTERN,
+  INTERPRETATION_TO_CODE,
+} from 'src/constants/fhir';
+
+/**
+ * Creates a FHIR Coding object
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: remove comments across files, if code is self-explanatory. If there is a complex logic - then retain comments

File: package.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
     "@babel/eslint-parser": "^7.28.4",
     "@babel/preset-env": "^7.28.3",
     "@babel/preset-react": "^7.27.1",
+    "@fortawesome/fontawesome-free": "^7.1.0",
```
Comment (sumaztwcode) [inline package.json:40]: need for this as part of this card?

File: src/helpers/FhirObservationTransformer.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+  const value = observationPayload.value;
+  const conceptDatatype =
+    typeof observationPayload.concept === 'object'
+      ? observationPayload.concept.datatype
+      : undefined;
+
+  if (value !== null && value !== undefined) {
+    switch (typeof value) {
+      case 'number':
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: move to constants - cases

File: src/helpers/FhirObservationTransformer.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+ * });
+ *
+ * @param {Array} observations - Raw observations from Container.getValue() or transformed Form2Observation[]
+ * @param {Object} options - Configuration options
+ * @param {Object} options.patientReference - FHIR Reference to patient (e.g., { reference: 'Patient/uuid' })
+ * @param {Object} options.encounterReference - FHIR Reference to encounter (e.g., { reference: 'Encounter/uuid' })
+ * @param {Object} options.performerReference - FHIR Reference to performer (e.g., { reference: 'Practitioner/uuid' })
+ * @returns {Array<{resource: Object, fullUrl: string}>} Array of FHIR Observation bundle entries
+ */
+export function transformToFhir(observations, options) {
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: is there a possibility that this function may throw error? If yes, how would it be conveyed to the consumer app?

File: src/helpers/FhirObservationTransformer.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+          observation.valueCodeableConcept = createCodeableConcept([
+            createCoding(value.uuid, undefined, value.display || value.displayString),
+          ]);
+        }
+        break;
+    }
+  }
+
+  if (observationPayload.interpretation) {
+    const interpretationValue = observationPayload.interpretation.toUpperCase();
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: blocking: This will throw TypeError if `interpretation` is not a string (e.g., null, number, object).

Suggested fix:
```javascript
if (observationPayload.interpretation) {
  const interpretationString = String(observationPayload.interpretation);
  const interpretationValue = interpretationString.toUpperCase();
  // ...
}
```

File: src/helpers/FhirObservationTransformer.js (hunk 5 above)
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: suggestion: Consider validating required `options` fields. Currently if a consumer calls this without patientReference/encounterReference/performerReference, it creates FHIR observations with undefined references.

```javascript
if (!options || !options.patientReference || !options.encounterReference || !options.performerReference) {
  throw new Error('transformToFhir requires patientReference, encounterReference, and performerReference in options');
}
```

File: src/helpers/FhirObservationTransformer.js [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        } else {
+          handleStringValue(value, observation, conceptDatatype);
+        }
+        break;
+      }
+      case 'boolean':
+        observation.valueBoolean = value;
+        break;
+      case 'object':
+        if (value instanceof Date) {
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js]: nit: Invalid Date objects will throw on `toISOString()`. Consider:

```javascript
if (value instanceof Date && !isNaN(value.getTime())) {
  observation.valueDateTime = value.toISOString();
}
```

File: src/helpers/FhirObservationTransformer.js [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+/**
+ * Generates a UUID v4
+ * @returns {string} UUID string
+ */
+const generateUUID = () => {
+  if (typeof crypto !== 'undefined' && crypto.randomUUID) {
+    return crypto.randomUUID();
+  }
+  // Fallback for environments without crypto.randomUUID
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
```
Comment (sumaztwcode) [inline src/helpers/FhirObservationTransformer.js:53]: nit: The fallback uses `Math.random()` which isn't cryptographically secure. Consider using `crypto.getRandomValues` as intermediate fallback, or at minimum log a warning when using Math.random() for healthcare data.

File: src/index.jsx [hunk 9]
```diff
@@ -64,6 +64,11 @@ export { BooleanValueMapper } from 'src/mapper/BooleanValueMapper';
 export { CodedValueMapper } from 'src/mapper/CodedValueMapper';
 export { CodedMultiSelectValueMapper } from 'src/mapper/CodedMultiSelectValueMapper';
 
+// -------------------------- FHIR transformation ---------------------
+export { transformToFhir } from 'src/helpers/FhirObservationTransformer';
```
Comment (sumaztwcode) [inline src/index.jsx]: nit: Both exports point to the same function (transformToFhir is the default export). Consider keeping only the named export `transformToFhir` to match existing patterns in this file.

File: src/index.jsx [hunk 10]
```diff
@@ -64,6 +64,11 @@ export { BooleanValueMapper } from 'src/mapper/BooleanValueMapper';
...
+export { default as FhirObservationTransformer } from 'src/helpers/FhirObservationTransformer';
+export * from 'src/constants/fhir';
```
Comment (sumaztwcode) [inline src/index.jsx]: nit: This differs from existing pattern where constants from `src/constants.js` are not exported from index.jsx. Consider grouping as a named export instead:
```javascript
import * as FhirConstants from 'src/constants/fhir';
export { FhirConstants };
```

**Review by sumaztwcode — COMMENTED** [review]
| File:Line | Type | Issue |
|-----------|------|-------|
| index.d.ts:384 | suggestion | `FhirObservationTransformer` export has no type annotation - either add proper typing or remove if `transformToFhir` function is sufficient |

File: package.json (hunk 1 above)
Comment (Aishwarya-thoughtworks) [reply package.json:71]: removed.

File: src/helpers/FhirObservationTransformer.js (hunk 2 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js]: removed the comments

File: package.json (hunk 3 above)
Comment (Aishwarya-thoughtworks) [reply package.json:40]: removed

File: src/helpers/FhirObservationTransformer.js (hunk 4 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js]: moved to constants

File: src/helpers/FhirObservationTransformer.js (hunk 5 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js]: done as suggested.

File: src/helpers/FhirObservationTransformer.js (hunk 7 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js]: done.

File: src/helpers/FhirObservationTransformer.js (hunk 8 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js:53]: Gave warning as crypto.randomValue() was not working.

File: src/index.jsx (hunk 9 above)
Comment (Aishwarya-thoughtworks) [reply src/index.jsx]: done

File: src/index.jsx (hunk 10 above)
Comment (Aishwarya-thoughtworks) [reply src/index.jsx]: done

File: src/helpers/FhirObservationTransformer.js (hunk 5 above)
Comment (Aishwarya-thoughtworks) [reply src/helpers/FhirObservationTransformer.js]: added as per the suggestion

**Review by sumaztwcode — APPROVED** [review]
In the commit messages, follow a standard - I see commits with no branch details, [v0.0.7]etc

File: src/helpers/FhirObservationTransformer.js
Comment (PratitiP) [inline src/helpers/FhirObservationTransformer.js:1]: if we can remove examples and detailed comments from fhirObservationTransformer.js file that will keep the code cleaner. We can add a ADR or add it to some document in detail. Each method meaning, what transformation it is doing and explain with example. So detailed example will be very easy to understand the FHIR transformation
